### ISSUE TITLE
Replace dyna_tools.py with lily-parsekit tools

### ISSUE THREAD

**Issue by FascinatedBox** [FascinatedBox/lily#302]
The way that Lily loads C libraries is different than how many other interpreters do it. Many interpreters will have a glue library and within it a special 'boot' function which is loaded and which tells the interpreter what the exports are. Lily doesn't do that, because there's a fair amount of overhead that comes from simply holding static type information.

Instead, Lily requires that glue libraries provide strings detailing the contents, but not to initialize them in case they aren't needed. This is used somewhat in the builtin libraries too, as a means of cutting down on the overhead of the builtin types and methods. This process has thus far been difficult at best, as special comment strings need to be sprinkled throughout a glue library's C source and then dyna_tools.py needs to be called whenever there is a library change. This is because, in addition to generating the dynaload string table, dyna_tools.py also generates helper macros used for creating new types, getting ids, which are stored in different files.

The current strategy, as I see it, has the following issues:

* Getting started is difficult. If I want to create a class, I need to update the file with the class, stop and refresh the bindings so I get the macros that I need.

* It's difficult to see at a glance what a library exports.

* Currently, modules like my apache bridge work by having the dynaload table in the same file as the embedder. This means that the dynaload table gets stuck in the middle somewhere.

* Helper macros are generated in a different file.

* Generates 1-2 files per source file. One for dyna, and one for extras if you need them.

What I'm proposing is to rewrite dyna_tools.py into a pure Lily script. This new script will likely be the basis of more tools to be coming in the (hopefully near) future. I have a tentative name in mind of 'petal' for the script. What I want to do is a new system that works as follows:

First, that for any .c file you plan to export, you have a `.frame.lily` (tentative name here too) file. The aim of a frame file is that it looks like what you plan to be exporting. So for the builtin library I might write:

```
### Provides a wrapper around C's stdin.
var stdin: File
var stdout: File

### Lily's basic numeric class. Wraps over a C int64_t.
foreign class Integer {
    ### Convert an `Integer` to a `String`.
    define to_s: String
}

native class Exception {
    var @message: String
    var @traceback: List[String]
}
```

I think that when someone plans to create a bridge from Lily to some C library, they have a rough idea of what the library API is, and what Lily provides. So that someone can create one of these frame files, and give it docblock comments as they go along. Once they're done planning the framework, they can order the tool to generate (or refresh if it already exists) a corresponding C file.

Once petal (or whatever it ends up being called) is ready, I'll create a script that will extract the dyna_tools comments out and make starter framework files. I'll host that tool in a different repo and leave it up for a while in case anyone needs it and happens to be late to the party.

I also want to begin a convention wherein an embedder of the language should not exist in the same file as a library it's embedded. So, in the case of my apache library, the `server` source and the driver (the apache hook as well as the rest) would be in two different files.

I think it's also a good idea to instruct users in documentation that they should have their frame file created before they begin to do any binding work. I plan for the tool to place macros in a block that's delimited by a special marker (ex: `/** begin Lily macros **/` ... `/** end Lily macros **/`). Since embedders won't be in the same file as a library, it's therefore easy to place the dynaload table at the very end of a file (and if necessary, wrapping it in comment markers as well). If a loader is needed, then it will be placed behind the dynaload table and be visible.

This tool will assume that binding libraries are self-contained. That is, the interpreter should not need to do any init or deinit work within the C side of the interpreter. That work should be handled through the Lily side, during vm execution. As such, it's assumed the `.c` file won't need a corresponding `.h` file because the only exports would be a dynaload table and a loader. Lily's api may be extended to provide an easy means for an embedder to declare and make use of libraries without generating a header. For example:

```
/* This would write externs for server parts. */
LILY_EMBED_LIBRARY(server)

{
    lily_state *s = lily_new_state(...);
    /* Registers the 'server' import to Lily. */
    LILY_REGISTER_LIBRARY(s, server)
    lily_parse_file(s, ...);
}
```

Part of the transition will involve changing how dynaload tables are written down. Currently, each library writes down `lily_dynaload_table`, and that's what the interpreter looks for. For a short time, I'll have it look for either that, or `lily_${LIBRARY_NAME}_dynaload_table`. Once I feel enough time has elapsed for a proper migration, I'll switch it to only look for namespaced tables and loaders.

Currently, dyna_tools.py is also responsible for generating (ugly) documentation from what it scans. As part of the transition, I'll create a wrapper over CommonMark (preferred) or a C markdown library for Lily. This will allow the documentation that's already written to be kept.

**Comment by FascinatedBox** [comment]
I started this a couple of weeks ago but forgot to assign myself to the bug. So far, the project I have is roughly 2K lines of pure Lily. In writing it, I uncovered a few bugs, and have gotten to appreciate a few features in the language as well. It's by far the biggest project ever made in the language. When the last part of this is done, I'll post a writeup about it.

What I decided to make was a system very close to the original. Same special comments, though slightly modified. Enough of a change that the old `dyna_tools.py` can't read the comments now, which means the super bad documentation can't be built.

The big difference is that the new tool autogenerates data into a section that's marked by the user with a different kind of special comment. So no more dyna_ or extras_ files. Everything gets a loader though, which is a bit more expensive where the parser used to infer that information.

Now instead of regexps for processing, this is using an engine that understands enough of Lily's syntax to process what seems to be everything I currently use.

The current plan is to:

* When time permits, running through all of the libraries that I built to migrate them over.
* Using the existing engine, layer a documentation builder on top that spits out 1 file per class/enum, and some html styling.
* Regenerating all current documentation, discarding the old (it's bad), and full deletion of dyna_tools.py
